gulp-tasks/test.js

```diff
@@ -1,47 +1,69 @@
 const gulp = require('gulp');
 const oneLine = require('common-tags').oneLine;
+const glob = require('glob');
+const fse = require('fs-extra');
+const path = require('path');
 
 const spawn = require('./utils/spawn-promise-wrapper');
 const getNpmCmd = require('./utils/get-npm-cmd');
 const logHelper = require('../infra/utils/log-helper');
 
-const runNodeTestsWithEnv = (nodeEnv) => {
-  return () => {
-    logHelper.log(oneLine`
-      Running Node Tests.
-    `);
-    const options = [];
-    if (global.cliOptions.grep) {
-      options.push('--grep', global.cliOptions.grep);
-    }
-    const originalNodeEnv = process.env.NODE_ENV;
-
-    process.env.NODE_ENV = nodeEnv;
-    return spawn(getNpmCmd(), ['run', 'test', '--',
-      `./test/${global.packageOrStar}/node/**/*.{js,mjs}`,
+const runNodeTestSuite = async (testPath, nodeEnv) => {
+  logHelper.log(oneLine`
+  Running Node test on ${logHelper.highlight(testPath)}
+  with NODE_ENV '${nodeEnv}'
+`);
+
+  const options = [];
+  if (global.cliOptions.grep) {
+    options.push('--grep', global.cliOptions.grep);
+  }
+  const originalNodeEnv = process.env.NODE_ENV;
+
+  process.env.NODE_ENV = nodeEnv;
+  try {
+    await spawn(getNpmCmd(), ['run', 'test', '--',
+      `${testPath}/node/**/*.{js,mjs}`,
       ...options,
-    ])
-    .catch((err) => {
-      logHelper.error(err);
-      throw new Error(`[Workbox Error Msg] 'gulp test' discovered errors.`);
-    })
-    .then(() => {
-      process.env.NODE_ENV = originalNodeEnv;
-    });
-  };
+    ]);
+    process.env.NODE_ENV = originalNodeEnv;
+  } catch (err) {
+    process.env.NODE_ENV = originalNodeEnv;
+
+    logHelper.error(err);
+    throw new Error(`[Workbox Error Msg] 'gulp test' discovered errors.`);
+  }
+};
+
+const runNodeTestsWithEnv = (nodeEnv) => {
+  const packagesToTest = glob.sync(`test/${global.packageOrStar}`);
+  return packagesToTest.reduce(async (promiseChain, packageToTest) => {
+    await promiseChain;
+    return runNodeTestSuite(packageToTest, nodeEnv);
+  }, Promise.resolve());
 };
 
 gulp.task('test:node:prod', gulp.series(
-  runNodeTestsWithEnv('production'),
+  () => runNodeTestsWithEnv('production'),
 ));
 
 gulp.task('test:node:dev', gulp.series(
-  runNodeTestsWithEnv('dev'),
+  () => runNodeTestsWithEnv('dev'),
 ));
 
+gulp.task('test:node:clean', () => {
+  return fse.remove(path.join(__dirname, '..', '.nyc_output'));
+});
+
+gulp.task('test:node:coverage', () => {
+  return spawn(getNpmCmd(), ['run', 'coverage-report']);
```

Comment:
When reading up on `gulp` v4, I learned that there is native support for using a `spawn`ed process as a return value from a task: https://www.joezimjs.com/javascript/complete-guide-upgrading-gulp-4/#4-Return-a-Child-Process

I think we need to keep `./utils/spawn-promise-wrapper` around for use cases where there are multiple subprocesses, or cleanup needs to be done after the subprocess runs, but it should be usable here.

Reply:
Good to know.